Convert more of the media code to async/await

clokep opened the pull request:
So this converts *most* of the remaining media code to async/await. The remaining pieces after this are the `StorageProvider` and sub-classes, but I think those need changes to the S3 code at the same time? I'm not really sure about that.

I've held off on pushing these changes for a while since the code surrounding them is confusing and I was not super confident. Please review thoroughly (and also consider if any of this will affect the S3 provider or other third-party providers).

richvdh commented on synapse/rest/media/v1/media_storage.py:
```diff
@@ -46,8 +46,7 @@ def __init__(self, hs, local_media_directory, filepaths, storage_providers):
 
-    @defer.inlineCallbacks
-    def store_file(self, source, file_info):
+    async def store_file(self, source, file_info):
```
docstring needs an update: no longer returns deferred.

richvdh commented on synapse/rest/media/v1/media_storage.py:
```diff
@@ -103,10 +102,9 @@ def store_into_file(self, file_info):
+        async def finish():
             for provider in self.storage_providers:
-                yield provider.store_file(path, file_info)
+                await provider.store_file(path, file_info)
```
sadly I don't think we can rely on the storage providers honouring their interface. In particular, see https://github.com/matrix-org/synapse-s3-storage-provider/blob/master/s3_storage_provider.py#L96-L98: `reactor.callInThread` returns None, and `make_deferred_yieldable` passes it through.

in other words: better add some safety checks here.

clokep replied:
I added some checks in af4f079bf5ae5bec16868da61c4858b1c68bf672 for this, hopefully they make sense!

richvdh commented:
> The remaining pieces after this are the StorageProvider and sub-classes, but I think those need changes to the S3 code at the same time?

that shouldn't be needed. Basically if we make sure that wherever we get a result from `StorageProvider.{fetch,store_file}` we handle either Deferreds or Coroutines, then we can change the `StorageProvider` interface to say that the methods return `Awaitables` instead, and the S3 storage provider will still be compatible. Then we can update the `StorageProvider` implementations to be async.

clokep commented:
Thanks! :) I'll take another look and see how I can add that isolation in.

richvdh commented on synapse/rest/media/v1/media_storage.py:
```diff
@@ -104,7 +105,11 @@ def store_into_file(self, file_info):
+                # store_file is supposed to return an Awaitable or Deferred, but
+                # guard against improper implementations.
+                result = provider.store_file(path, file_info)
+                if inspect.isawaitable(result) or isinstance(result, defer.Deferred):
```
I think Deferred is awaitable.

clokep replied:
Oh, right. That's why using `await` with it works! 🤦

clokep commented:
> I think you could merge this as-is (well, after fixing the issues) and do the rest of StorageProvider as another step.

I converted the other couple of methods here (the ones that use `StorageProvider.fetch` as well in this PR. I'll tackle the `StorageProvider` class itself in a separate PR, I think. I added the same guards as for `store_file`, although the S3 storage provider seems to be properly implemented in this case.
